src/msg/simple/Pipe.cc

```diff
@@ -2601,10 +2601,21 @@ int Pipe::tcp_read_wait()
 
 ssize_t Pipe::do_recv(char *buf, size_t len, int flags)
 {
+  int retries = 0;
 again:
   ssize_t got = ::recv( sd, buf, len, flags );
   if (got < 0) {
-    if (errno == EAGAIN || errno == EINTR) {
+    if (errno == EINTR) {
+      goto again;
+    } else if (errno == EAGAIN &&
+	       msgr->cct->_conf->ms_socket_retry_on_eagain > retries) {
```

Comment:
You've identified one circumstance under which we get EAGAIN (hey, on Linux it might even be the only one) but unless we can disprove the existence of others, we need to account for them. This loop isn't supposed to handle the case where we really did "lose" data we thought was available.